submission/races.go

```diff
@@ -0,0 +1,299 @@
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sct == nil {
+		sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		return
+	}
+	// If at least one group needs that SCT, result is set. Otherwise ignored.
+	for groupName := range sub.logToGroups[logURL] {
+		if sub.groupNeeds[groupName] > 0 {
+			sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		}
+		sub.groupNeeds[groupName]--
+	}
+
+	// Cancel any pending Log-requests for which there's no more awaiting Log-group.
+	for logURL, groupSet := range sub.logToGroups {
+		isAwaitedByGroup := false
+		for g := range groupSet {
+			if sub.groupNeeds[g] > 0 {
+				isAwaitedByGroup = true
+				break
+			}
+		}
+		if !isAwaitedByGroup && sub.cancels[logURL] != nil {
+			sub.cancels[logURL]()
+			delete(sub.cancels, logURL)
+		}
+	}
+}
+
+// returns whether the group has stopped waiting for more SCTs.
+func (sub *safeSubmissionState) groupComplete(groupName string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	needs, ok := sub.groupNeeds[groupName]
+	if !ok {
+		return true
+	}
+	return needs <= 0
+}
+
+func (sub *safeSubmissionState) collectSCTs() []*AssignedSCT {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	scts := []*AssignedSCT{}
+	for logURL, r := range sub.results {
+		if r != nil && r.sct != nil {
+			scts = append(scts, &AssignedSCT{LogURL: logURL, SCT: r.sct})
+		}
+	}
+	return scts
+}
+
+// submitInterval calculates duration for consequent call.
+// For first parallelStart calls duration is 0, while every next one gets
+// additional dur interval.
+func submitInterval(idx int, parallelStart int, dur time.Duration) time.Duration {
+	if idx < parallelStart {
+		return time.Duration(0)
+	}
+	return time.Duration(idx+1-parallelStart) * dur
+}
+
+// Submitter is interface wrapping Log-request-response cycle and any processing.
+type Submitter interface {
+	SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error)
+}
+
+// groupRace shuffles logs within the group, submits avoiding duplicate-requests
+// and collects responses.
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	state *safeSubmissionState, submitter Submitter) groupState {
+	groupURLs := make([]string, 0, len(group.LogURLs))
+	for logURL := range group.LogURLs {
+		groupURLs = append(groupURLs, logURL)
+	}
+
+	type count struct{}
+	counter := make(chan count, len(groupURLs))
+
+	// Randomize the order in which we send requests to the logs in a group
+	// so we maximize the distribution of logs we get SCTs from.
+	for i, urlNum := range rand.Perm(len(groupURLs)) {
+		subCtx, cancel := context.WithCancel(ctx)
+		go func(i int, logURL string) {
+			timeoutchan := time.After(submitInterval(i, parallelStart, SubmitBatchInterval))
+			select {
+			case <-subCtx.Done():
+				counter <- count{}
+				return
+			case <-timeoutchan:
+			}
+			if state.groupComplete(group.Name) {
```

Comment:
Have you considered making the `groupComplete` indicator a channel that closes when the group is done, similar to `Context.Done`? If it was the case, you could include the `groupComplete` channel into the `select` above.

Reply:
That "group done" channel could be also derived from some context, or even be a context if you bound the lifetime of the group state by the per-group request context.

Reply:
The main problem with channel is: you cannot find out whether it's open or not until you try writing into that.
And we need this info (open/closed or incomplete/complete) long before we're able to write anything into that.

Reply:
I think that in this case it's not necessary to store the `groupComplete` bit outside of `groupRace` function, and all the reactiveness can boil down to:

- In the beginning of `groupRace` create a cancelable context: `cctx, cancel := context.WithCancel(ctx)`
- Derive `subCtx`s from `cctx`
- `cancel()` the `cctx` when you have sufficient number of SCTs
- This will kill all `subCtx`s and the corresponding goroutines automatically

Reply:
The major problem with this approach: even when one group has collected enough SCTs we cannot just stop it's context. As some its requested-but-not-yet-responded Logs may be crucial for other groups. Due to Log-groups intersection.

Reply:
Yep, I'm not telling to cancel the log requests, just cancel everything that runs in `groupRace` function, i.e. goroutines and stuff. You don't have to derive log requests' context from the group context if you want it to be independent. A solution to that is having 2 "root" contexts - one for groups, another one for individual log requests.

Reply:
Having said that, I would continue my point listed in some other comments: it probably makes sense to not overload this PR with optimizations. Not stopping log requests and trying to synchronize all groups w.r.t. having only one request per log definitely sounds like an optimization that deserves a separate PR. To submit this PR, it would be enough to cancel the whole group with all the requests.

For example, as a start, is it a reasonable assumption to make that the groups are not intersecting each other (except the base group)? Non-Google logs vs. Google logs don't intersect, so what's the point of synchronizing them this way? On top of that, you already random shuffle the logs in the group, so the probability of multiple groups hitting the same logs after shuffle is low, even if the groups intersect.

Reply:
Another observation is: CT logs can handle duplicate submissions. Even if multiple groups try to submit the same cert to the same log, they are likely to get the same SCT.